connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java

```diff
@@ -701,43 +745,151 @@ protected void produceMessages(EmbeddedConnectCluster cluster, String topicName,
+                                "Consumer group committed downstream offsets beyond the log end, this would lead to negative lag metrics"
+                        );
+                    }
+                }
                 boolean totalOffsetsMatch = exactOffsetTranslation
-                        ? totalOffsets == expectedTotalOffsets
-                        : totalOffsets >= expectedTotalOffsets;
+                        ? totalEndOffsets == expectedTotalOffsets
+                        : totalEndOffsets >= expectedTotalOffsets;
+
+                boolean consumerGroupOffsetsMatch = exactOffsetTranslation
+                        ? totalConsumerGroupOffsets == expectedTotalOffsets
+                        : totalConsumerGroupOffsets >= expectedTotalOffsets;
                 // make sure the consumer group offsets are synced to expected number
-                return totalOffsetsMatch && consumerGroupOffsetTotal > 0;
+                return totalOffsetsMatch && consumerGroupOffsetsMatch;
             }, OFFSET_SYNC_DURATION_MS, "Consumer group offset sync is not complete in time");
         }
     }
 
+    protected static void insertDummyOffsetSyncs(EmbeddedConnectCluster cluster, String offsetSyncsTopic, String topic, int partitions) throws InterruptedException {
+        waitForTopicCreated(cluster, offsetSyncsTopic);
+        // Insert a large number of checkpoint records into the offset syncs topic to simulate
+        // a long-lived MM2 instance that has replicated many offsets in the past.
+        List<ProducerRecord<byte[], byte[]>> records = new ArrayList<>();
+        for (int partition = 0; partition < partitions; partition++) {
+            TopicPartition tp = new TopicPartition(topic, partition);
+            OffsetSync sync = new OffsetSync(tp, 0L, 0L);
+            records.add(new ProducerRecord<>(offsetSyncsTopic, sync.recordKey(), sync.recordValue()));
+        }
+        Map<String, Object> producerProps = new HashMap<>();
+        int sentRecords = 0;
+        try (KafkaProducer<byte[], byte[]> producer = cluster.kafka().createProducer(producerProps)) {
+            // Try to ensure that the contents of the offset syncs topic is too large to read before the checkpoint
+            // interval passes, so that the first checkpoint would take place before reading the whole contents of the
+            // sync topic. Experimentally, this test passes with <2x, and fails with >5x, without a fix for KAFKA-13659.
+            double consumeEfficiency = 10;
+            long deadline = System.currentTimeMillis() + (long) (consumeEfficiency * CHECKPOINT_INTERVAL_DURATION_MS);
+            int nRecords = records.size();
+            while (System.currentTimeMillis() < deadline) {
+                producer.send(records.get(sentRecords % nRecords));
+                sentRecords++;
+            }
+            producer.flush();
+        }
+        log.info("Sent {} dummy records to {}", sentRecords, offsetSyncsTopic);
+    }
```

Comment:
This is interesting, but seems a little too prone to flakiness to be worth the benefits it provides.

IMO we can remove this method (and possibly the entire `testRestartReplication` test case) in favor of adding unit testing coverage for the `MirrorCheckpointTask` and/or `OffsetSyncStore` classes.

Reply:
I'll remove insertDummyOffsetSyncs, but leave the testRestartReplication, as it is still testing a new situation: restarting the connectors mid-replication.
It won't target this specific bug without the `insertDummyOffsetSyncs`, but could still add value.

I've added unit tests for OffsetSyncStore that address the bug (translating before start finishes) that should hopefully compensate for this test's smaller scope.

I looked into testing MirrorCheckpointTask, and it looks too involved to address in this PR. I think it would be more appropriate to follow up later with some refactoring & testing for the checkpoint task.
I feel confident from the results of the insertDummyOffsetSyncs test that:
1. We are addressing the situation where the offset syncs topic is very large
2. We are not introducing a severe performance regression, as the consumeEfficiency is very high.